x-pack/functionbeat/provider/aws/kinesis.go

```diff
@@ -17,14 +22,109 @@ import (
 	"github.com/elastic/beats/x-pack/functionbeat/provider/aws/transformer"
 )
 
+type startingPosition uint
+
+const (
+	// Looking at the documentation, Kinesis should also support `AT_TIMESTAMP` but looking at the
+	// request format for cloudformation, I don't see a way to define the timestamp.
+	// I've looked at other frameworks, and it seems a bug in the cloudformation API.
+	// doc: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html
+	trimHorizonPos startingPosition = iota + 1
+	latestPos
+)
+
+var (
+	mapStartingPosition = map[string]startingPosition{
+		"trim_horizon": trimHorizonPos,
+		"latest":       latestPos,
+	}
+
+	mapStartingPositionReverse = make(map[startingPosition]string, len(mapStartingPosition))
+)
+
+func init() {
+	for k, v := range mapStartingPosition {
+		mapStartingPositionReverse[v] = strings.ToUpper(k)
+	}
+}
+
+func (s *startingPosition) Unpack(str string) error {
+	v, ok := mapStartingPosition[str]
+	if !ok {
+		validValues := make([]string, len(mapStartingPosition))
+		pos := 0
+		for k := range mapStartingPosition {
+			validValues[pos] = k
+			pos++
+		}
+		return fmt.Errorf("unknown value %s, valid values are: %s", str, strings.Join(validValues, ", "))
+	}
+	*s = v
+	return nil
+}
+
+func (s *startingPosition) String() string {
+	v, ok := mapStartingPositionReverse[*s]
+	if !ok {
+		panic("unknown starting position: " + string(*s))
+	}
+	return v
+}
+
+// KinesisConfig is the configuration for the Kinesis event type.
+type KinesisConfig struct {
+	Description  string                  `config:"description"`
+	Name         string                  `config:"name" validate:"nonzero,required"`
+	Triggers     []*KinesisTriggerConfig `config:"triggers"`
+	LambdaConfig *lambdaConfig           `config:",inline"`
+}
+
+// Validate validates the configuration.
+func (cfg *KinesisConfig) Validate() error {
+	if len(cfg.Triggers) == 0 {
+		return errors.New("you need to specify at least one trigger")
+	}
+	return nil
+}
+
+// KinesisTriggerConfig configuration for the current trigger.
+type KinesisTriggerConfig struct {
+	EventSourceArn   string
+	BatchSize        int
+	StartingPosition startingPosition
+}
+
+// Unpack unpacks the trigger and make sure the defaults settings are correctly sets.
+func (c *KinesisTriggerConfig) Unpack(cfg *common.Config) error {
+	// Create an anonymous struct so we don't go into a recursive unpack.
+	config := struct {
+		EventSourceArn   string           `config:"event_source_arn" validate:"required"`
+		BatchSize        int              `config:"batch_size" validate:"min=100,max=10000"`
+		StartingPosition startingPosition `config:"starting_position"`
+	}{
+		BatchSize:        100,
+		StartingPosition: trimHorizonPos,
+	}
```

Comment:
I will change the code to do this instead `type XX KinesistriggerConfig` and use XX to define the default and leave the tags on the main struct.